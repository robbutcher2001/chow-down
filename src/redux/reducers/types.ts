export interface Message {
    user: string
    message: string
}

export interface HelloState {
    hello_messages: Message[]
}

export const SAY_HELLO = 'SAY_HELLO'
export const DELETE_HELLO = 'DELETE_HELLO';

interface SayHelloAction {
    type: typeof SAY_HELLO
    rob_payload: Message
}

interface DeleteHelloAction {
    type: typeof DELETE_HELLO
}

export type SayHelloTypes = SayHelloAction | DeleteHelloAction