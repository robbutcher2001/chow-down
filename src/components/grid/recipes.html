<!DOCTYPE html>
<html>

<head>
    <title>Grid</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="recipes.css">
</head>

<body>
    <nav>
        <button>Home</button>
        <button class="active">Recipes</button>
        <button>Settings..</button>
    </nav>
    <header>
        <h1>Chow Down</h1>
        <p>Chow down on a weekly plan of your evening meals.</p>
    </header>
    <main>
        <div>
            <h2 class="highlighter">Your recipes</h2>
            <section>
                <img src="eclipse.svg"></img>
            </section>
            <ul>
                <li class="highlighter">
                    <figure>
                        <img src="https://www.bbcgoodfood.com/sites/default/files/styles/recipe/public/recipe/recipe-image/2017/01/jacket-potato.jpg?itok=jFmhmfQn"
                            alt="Some recipe">
                        <figcaption>
                            <h3>Jacket potato with whipped feta & sumac</h3>
                        </figcaption>
                    </figure>
                    <section>
                        <mark>&#9733;</mark>
                        <mark>&#9733;</mark>
                        <mark>&#9733;</mark>
                    </section>
                    <p>The perfect budget-friendly, filling supper for one. The fragrant, zesty flavour of sumac is a
                        refreshing contrast to the creamy whipped feta.</p>
                </li>
                <li class="highlighter">
                    <figure>
                        <img src="https://www.bbcgoodfood.com/sites/default/files/styles/recipe/public/recipe/recipe-image/2016/08/curried-cod.jpg?itok=7P5xS7qS"
                            alt="Some recipe">
                        <figcaption>
                            <h3>Curried cod</h3>
                        </figcaption>
                    </figure>
                    <section>
                        <mark>&#9733;</mark>
                        <mark>&#9733;</mark>
                    </section>
                    <p>An easy-to-prepare midweek one-pot with cod fillet, chickpeas, ginger and spices - it's healthy,
                        low calorie and packed with iron too!</p>
                </li>
                <li class="highlighter">
                    <figure>
                        <img src="https://www.bbcgoodfood.com/sites/default/files/styles/recipe/public/recipe/recipe-image/2016/07/peanut-butter-chicken.jpg?itok=mg1THxCG"
                            alt="Some recipe">
                        <figcaption>
                            <h3>Peanut butter chicken</h3>
                        </figcaption>
                    </figure>
                    <section>
                        <mark>&#9733;</mark>
                        <mark>&#9733;</mark>
                        <mark>&#9733;</mark>
                        <mark>&#9733;</mark>
                        <mark>&#9733;</mark>
                    </section>
                    <p>The whole family will love this new, budget chicken dish. Any leftovers freeze well and make a
                        handy lunch.</p>
                </li>
                <li class="highlighter">
                    <figure>
                        <img src="https://www.bbcgoodfood.com/sites/default/files/styles/recipe/public/user-collections/my-colelction-image/2015/12/recipe-image-legacy-id--513635_11.jpg?itok=CpUnFxL4"
                            alt="Some recipe">
                        <figcaption>
                            <h3>Spring chicken in a pot</h3>
                        </figcaption>
                    </figure>
                    <section>
                        <mark>&#9733;</mark>
                        <mark>&#9733;</mark>
                        <mark>&#9733;</mark>
                        <mark>&#9733;</mark>
                    </section>
                    <p>Casseroles aren't just for winter – this light, vibrant one-pot is packed with spring veg and
                        herby pesto.</p>
                </li>
                <li class="highlighter">
                    <figure>
                        <img src="https://www.bbcgoodfood.com/sites/default/files/styles/recipe/public/recipe/recipe-image/2017/09/oregano-chicken-squash-traybake.jpg?itok=lX0GQDHS"
                            alt="Some recipe">
                        <figcaption>
                            <h3>Oregano chicken & squash traybake</h3>
                        </figcaption>
                    </figure>
                    <section>
                        <mark>&#9733;</mark>
                    </section>
                    <p>Try this low-calorie chicken, butternut squash and artichoke traybake as an easy dinner during
                        busy weekdays. It takes just five minutes to prepare.</p>
                </li>
                <li class="highlighter">
                    <figure>
                        <img src="https://www.bbcgoodfood.com/sites/default/files/styles/recipe/public/recipe/recipe-image/2018/07/chilli-garlic-oregano-spatchcock-chicken-with-chipotle-aioli.jpg?itok=oGn3azLS"
                            alt="Some recipe">
                        <figcaption>
                            <h3>Chilli, garlic & oregano spatchcock chicken with chipotle aïoli</h3>
                        </figcaption>
                    </figure>
                    <section>
                        <mark>&#9733;</mark>
                        <mark>&#9733;</mark>
                        <mark>&#9733;</mark>
                        <mark>&#9733;</mark>
                        <mark>&#9733;</mark>
                    </section>
                    <p>Take simple ingredients beyond the norm with this fabulous spicy, herby spatchcock chicken with
                        chipotle aïoli. Serve with sweet potato wedges and salad.</p>
                </li>
                <li class="highlighter">
                    <figure>
                        <img src="https://www.bbcgoodfood.com/sites/default/files/styles/recipe/public/recipe/recipe-image/2017/01/jacket-potato.jpg?itok=jFmhmfQn"
                            alt="Some recipe">
                        <figcaption>
                            <h3>Jacket potato with whipped feta & sumac</h3>
                        </figcaption>
                    </figure>
                    <section>
                        <mark>&#9733;</mark>
                        <mark>&#9733;</mark>
                        <mark>&#9733;</mark>
                    </section>
                    <p>The perfect budget-friendly, filling supper for one. The fragrant, zesty flavour of sumac is a
                        refreshing contrast to the creamy whipped feta.</p>
                </li>
                <li class="highlighter">
                    <figure>
                        <img src="https://www.bbcgoodfood.com/sites/default/files/styles/recipe/public/recipe/recipe-image/2016/08/curried-cod.jpg?itok=7P5xS7qS"
                            alt="Some recipe">
                        <figcaption>
                            <h3>Curried cod</h3>
                        </figcaption>
                    </figure>
                    <section>
                        <mark>&#9733;</mark>
                        <mark>&#9733;</mark>
                    </section>
                    <p>An easy-to-prepare midweek one-pot with cod fillet, chickpeas, ginger and spices - it's healthy,
                        low calorie and packed with iron too!</p>
                </li>
            </ul>
        </div>
    </main>
    <footer class="highlighter">
        <p>&copy; Rob Butcher 2019.</p>
        <p>Images: <a href="https://unsplash.com/" target="_blank">Unsplash</a> and recipe websites.</p>
    </footer>
</body>

</html>