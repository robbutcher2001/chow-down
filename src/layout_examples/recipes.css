/* refine and use fonts properly */
@import url('https://fonts.googleapis.com/css?family=Lato:400,400italic,700,700italic|Source+Code+Pro:400');

html {
    font-family: "Lato", sans-serif;
    height: 100%;
}

body {
    display: grid;
    align-items: flex-start;
    grid-template-rows: auto auto 1fr auto;
    min-height: 100%;
    margin: 0;
}

/* just for dev */
.highlighter2 {
    border-width: 1px;
    border-color: black;
    border-style: dotted;
}

nav {
    display: flex;
    flex-direction: row;
    justify-content: flex-start;
    position: -webkit-sticky;
    position: sticky;
    top: 0;
    z-index: 100;
    background: #222;
    padding: 0 .75rem;
}

nav>button {
    background: #222;
    font-size: 1rem;
    padding: 1rem 1rem 0.5rem 1rem;
    border: none;
    border-bottom: .5rem solid #222;
    color: #fff;
    cursor: pointer;
}

nav>button:hover, nav>button.active {
    border-bottom-color: #4acaa8;
}

nav>button:focus {
    outline: none;
}

header {
    display: grid;
    align-items: flex-end;
    grid-template-columns: repeat(auto-fit, minmax(280px, 1fr));
    gap: .5em;

    background-image: linear-gradient(to bottom, rgba(0, 0, 0, 0.4), rgba(0, 0, 0, 0.8)), url("dinner-1.jpg");
    background-position: center;
    background-repeat: no-repeat;
    background-size: cover;
    padding: 2rem 1rem;
    min-width: 280px; /* needs to be 280px + 2rem when using scss calculate */
}

header>* {
    color: white;
    padding: .75rem;
    margin: 0;
}

header>h1 {
    font-size: 3rem;
}

header>p {
    font-size: 1.5rem;
}

main {
    display: flex;
    justify-content: center;
    margin: 1rem 1rem 8rem 1rem;
}

main>div {
    max-width: 1100px;
}

main h2 {
    color: #4acaa8;
    font-size: 2rem;
    margin: 0;
}

main section {
    display: flex;
    justify-content: center;
    margin-top: 2rem;
}




/* copied up to here to Page component */


main ul {
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax(280px, 1fr));
    gap: 1rem 2em;
    padding: 1rem 0rem 0rem 0rem;
    margin: 0;
    list-style: none;
}

main ul>li {
    -webkit-box-shadow: 0 1px 0.25rem rgba(0, 0, 0, .08);
    box-shadow: 0 1px 0.25rem rgba(0, 0, 0, .08);
    border: 1px solid rgba(0,0,0,.04);
    border-radius: 0.5rem;
    cursor: pointer;
}

main ul>li>figure {
    max-height: 220px;
    overflow: hidden;
    margin: 0;
    border-top-left-radius: .5rem;
    border-top-right-radius: .5rem;
    position: relative;
}

main ul>li>figure>img {
    width: 100%;
}

main ul>li>figure>figcaption {
    position: absolute;
    bottom: 0;
    background-color: rgba(0, 0, 0, .7);
    width: 100%;
}

main ul>li>figure>figcaption>h3 {
    color: white;
    padding: .5rem;
    margin: 0;
}

main ul>li>p {
    color: #888;
    padding: 0 .5rem 1.5rem 0.5rem;
    margin: 0;
}

main ul>li>section {
    display: flex;
    flex-direction: row;
    justify-content: left;
    font-size: 1.2rem;
    padding: 0.5rem;
}

main ul>li>section>mark {
    color: #4acaa8;
    background-color: white;
}

footer {
    display: flex;
    flex-direction: column;
    justify-content: center;
    font-size: 0.8rem;
    padding: 3rem 0;
    color: #c0c0c0;
    background-color: #fafafa;
}

footer>p {
    margin: 1rem;
    text-align: center;
}