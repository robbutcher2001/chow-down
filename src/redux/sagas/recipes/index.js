'use strict';

import { all, takeLatest } from 'redux-saga/effects';

import getSaga from './get';
import postSaga from './post';

export default function* watcherSaga() {
    console.log('[saga] Started recipes saga');
    yield all([
        takeLatest('GET_RECIPES_REQUEST', getSaga),
        takeLatest('POST_RECIPE_REQUEST', postSaga)
    ]);
}