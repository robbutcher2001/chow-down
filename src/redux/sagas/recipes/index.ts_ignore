'use strict';

import { all, takeLatest } from 'redux-saga/effects';

import { Actions } from '../../../globals/constants';

import getSaga from './get';
import postSaga from './post';

export default function* watcherSaga() {
    console.log('[saga] Started recipes saga');
    yield all([
        takeLatest(Actions.recipes.GET_RECIPES_REQUEST, getSaga),
        takeLatest(Actions.recipes.POST_RECIPE_REQUEST, postSaga)
    ]);
}