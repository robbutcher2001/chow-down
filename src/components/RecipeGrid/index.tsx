import React from 'react';
import styled from 'styled-components';

import { Recipe } from '../../store/domain/recipes/types';
import RecipeCard from '../RecipeCard';

interface RecipeGridProps {
  recipes: Recipe[]
};

const RecipeGrid = styled.ul`
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(280px, 1fr));
  gap: 1rem 2em;
  padding: 1rem 0rem 0rem 0rem;
  margin: 0;
  list-style: none;
`

export default (props: RecipeGridProps) => (
  <RecipeGrid>
    {props.recipes.map((recipe, i) =>
      <RecipeCard
        key={i}
        title={recipe.title}
        description={recipe.description}
        rating={recipe.rating}
        imageUrl={recipe.image}
        imageAlt={recipe.title}
      />
    )
    }
  </RecipeGrid>
);