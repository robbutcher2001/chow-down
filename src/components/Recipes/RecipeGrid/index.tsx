import React from 'react';
import moment from 'moment';
import { Link } from 'react-router-dom';

import styled from 'styled-components';

import { Recipe } from '../../../store/domain/recipes/types';
import { Day, PutDayApiRequest } from '../../../store/domain/days/types';
import RecipeCard from '../RecipeCard';
import { NegativeBox } from '../../MessageBox';

interface RecipeGridProps {
  isLoading: boolean,
  recipes: Recipe[],
  selectedDay?: string,
  putDay?: (day: Day) => PutDayApiRequest
};

const UserInstruction = styled.div`
  margin: 1rem 0 2rem;
  font-size: 1.2rem;

  .link {
    border: none;
    background: none;
    margin: 1rem 0;
    padding: 0;
    font-family: 'Lato', sans-serif;
    font-weight: 700;
    color: #005ea5;
    cursor: pointer;
    text-decoration: underline;
  }
`

const RecipeGrid = styled.ul`
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(280px, 1fr));
  gap: 1rem 2em;
  margin: 0;
  padding: 1rem 0rem 0rem 0rem;
  list-style: none;
`

export default (props: RecipeGridProps) =>
  <>
    {props.selectedDay &&
      <UserInstruction>
        <span>Select for {moment(props.selectedDay).format('dddd')} or set as </span>
        <Link className='link' to={`/days/alternate/new?date=` + props.selectedDay} >
          alternate day
        </Link>
        <span>.</span>
      </UserInstruction>
    }
    {!props.isLoading && props.recipes.length === 0 ?
      <NegativeBox message='No recipes yet!' /> :
      <RecipeGrid className={props.isLoading ? 'spinner spinning' : 'spinner'} >
        {props.recipes.map((recipe, i) =>
          <RecipeCard
            key={i}
            recipe={recipe}
            selectedDay={props.selectedDay}
            putDay={props.putDay}
          />
        )}
      </RecipeGrid>
    }
  </>;