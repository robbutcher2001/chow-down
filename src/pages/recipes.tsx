import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';

import { GlobalState } from '../store';
import { Recipe, GetRecipesApiRequest, PostRecipeApiRequest } from '../store/domain/recipes/types';
import { getRecipesRequest, postRecipesRequest } from '../store/domain/recipes/actions';

import Loading from '../components/Loading';
import Page from '../components/page';
import RecipeGrid from '../components/RecipeGrid';
import RecipeCard from '../components/RecipeCard';
import Form from '../components/form';
import InputBox from '../components/input-box';

interface StateProps {
    error: string,
    failure: string,
    recipes: Recipe[],
    ui: {
        pending: {
            get: boolean,
            post: boolean
        }
    }
};

interface DispatchProps {
    getRecipes: () => GetRecipesApiRequest,
    postRecipe: (form: object) => PostRecipeApiRequest
};

interface OwnProps { };

interface OwnState { };

type CombinedProps = StateProps & DispatchProps & OwnProps;

class RecipesPage extends Component<CombinedProps, OwnState> {
    constructor(props: CombinedProps) {
        super(props);
    }

    requestRecipes = () => this.props.getRecipes();
    addRecipe = (form: object) => this.props.postRecipe(form);

    componentDidMount = () => this.props.getRecipes();

    createRecipeCards = (recipes: Recipe[]) =>
        recipes.map((recipe, i) =>
            <RecipeCard
                key={i}
                title={recipe.title}
                description={recipe.description}
                rating={recipe.rating}
                imageUrl={recipe.image}
                imageAlt={recipe.title}
            />
        );

    render = () => {
        console.log(this.props.recipes);

        if (this.props.ui.pending.get) {
            return (
                <Page title='Your recipes'>
                    <Loading />
                </Page>
            )
        };

        return (
            <Page title='Your recipes'>
                <RecipeGrid>
                    {this.createRecipeCards(this.props.recipes)}
                </RecipeGrid>
                <h4>List recipes</h4>
                <button onClick={this.requestRecipes}>
                    Press me to get all recipes
                    </button>
                {this.props.ui.pending.get &&
                    <div style={{ color: 'red' }}>Getting..</div>
                }
                <h2>Add a new recipe</h2>
                <Form
                    dispatch={this.addRecipe}
                    submitText='Add recipe'>
                    <InputBox
                        name='title'
                        placeholderText='Title'
                    />
                    <InputBox
                        name='description'
                        placeholderText='Description'
                    />
                    <InputBox
                        name='rating'
                        placeholderText='Rating'
                    />
                    <InputBox
                        name='url'
                        placeholderText='Url'
                    />
                    <InputBox
                        name='image'
                        placeholderText='Upload image'
                    />
                </Form>
                {this.props.ui.pending.post &&
                    <div>Adding your new recipe..</div>
                }
                {this.props.failure &&
                    <div style={{ color: 'red' }}>{this.props.failure}</div>
                }
                {this.props.error &&
                    <div style={{ color: 'red' }}>{this.props.error}</div>
                }
            </Page>
        );
    }
};

const mapStateToProps = ({ app, domain, ui }: GlobalState, ownProps: OwnProps): StateProps => ({
    error: app.error,
    failure: domain.recipe.failure,
    recipes: domain.recipe.recipes,
    ui: {
        pending: {
            get: ui.recipe.getPending,
            post: ui.recipe.postPending
        }
    }
});

const mapDispatchToProps = (dispatch: Dispatch, ownProps: OwnProps): DispatchProps => ({
    getRecipes: () => dispatch(getRecipesRequest()),
    postRecipe: (form: object) => dispatch(postRecipesRequest(form))
});

export default connect<StateProps, DispatchProps, OwnProps, GlobalState>
    (mapStateToProps, mapDispatchToProps)(RecipesPage);