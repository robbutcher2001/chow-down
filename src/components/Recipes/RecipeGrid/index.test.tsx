import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';

import RecipeGrid from '.';
import { Recipe } from '../../../store/domain/recipes/types';

const mockRecipes: Recipe[] = [{
  id: '123',
  title: 'fake_title1',
  description: 'fake_description1',
  rating: 4,
  url: 'fake_url1',
  image: 'fake_alt_text1'
},
{
  id: '124',
  title: 'fake_title2',
  description: 'fake_description2',
  rating: 3,
  url: 'fake_url2',
  image: 'fake_alt_text2'
}];

test('RecipeGrid basic snapshot render', () => {
  const recipeGrid = renderer.create(
    <RecipeGrid recipes={mockRecipes} />
  );

  expect(recipeGrid.toJSON()).toMatchSnapshot();
});

test('RecipeGrid correct tag content assertion', () => {
  const { getByText } = render(
    <RecipeGrid recipes={mockRecipes} />
  );

  expect(getByText(/fake_title1/).textContent).toEqual('fake_title1');
  expect(getByText(/fake_description1/).textContent).toEqual('fake_description1');
});

test('RecipeGrid render multiple recipes assertion', () => {
  const { getByText } = render(
    <RecipeGrid recipes={mockRecipes} />
  );

  expect(getByText(/fake_title2/).textContent).toEqual('fake_title2');
  expect(getByText(/fake_description2/).textContent).toEqual('fake_description2');
});