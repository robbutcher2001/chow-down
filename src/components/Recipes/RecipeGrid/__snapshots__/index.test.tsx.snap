// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeGrid basic snapshot render 1`] = `
<ul
  className="sc-ifAKCX lkqgzt spinner"
>
  <li
    className="sc-bwzfXH gWWprb"
  >
    <figure>
      <aside />
      <figcaption>
        <h3>
          fake_title1
        </h3>
      </figcaption>
    </figure>
    <section
      className="sc-bdVaJa dyQPNL"
    >
      <mark>
        ★
      </mark>
      <mark>
        ★
      </mark>
      <mark>
        ★
      </mark>
      <mark>
        ★
      </mark>
    </section>
    <p>
      fake_description1
    </p>
  </li>
  <li
    className="sc-bwzfXH heyaQo"
  >
    <figure>
      <aside />
      <figcaption>
        <h3>
          fake_title2
        </h3>
      </figcaption>
    </figure>
    <section
      className="sc-bdVaJa dyQPNL"
    >
      <mark>
        ★
      </mark>
      <mark>
        ★
      </mark>
      <mark>
        ★
      </mark>
    </section>
    <p>
      fake_description2
    </p>
  </li>
</ul>
`;
